### PR TITLE
Refactor and improve project API util methods

### LINKED ISSUE #27273: Need a project-api util function to get sourceRoot from the documentPath
**Description:**
Need a project-api util function to get **sourceRoot** from the **documentPath**. With the current APIs it always required to create a project instance to get the projectRoot. But we need to calculate sourceRoot **without creating a project instance** for example when maintaining a mapping of project instances against it's sourceRoot path.

```
{sourceRoot -> projectInstance}
```

**Steps to reproduce:**

**Affected Versions:**
2.0.0-SLP7

### LINKED ISSUE #27274: Need a project-api util function to get Document providing documentPath from Project
**Description:**
Need a project-api util function to get Document providing documentPath from Project.

**Affected Versions:**
2.0.0-SLP7

### PR DESCRIPTION
**Pull request by azinneera**
## Purpose

* Introduces the following methods:
_**io.ballerina.projects.directory.ProjectPaths**_

**Update:**

`public static Path packageRoot(Path filepath) throws ProjectException;`
`public static boolean isBalFile(Path filepath);`
`public static boolean isStandaloneBalFile(Path filepath);`

~~`public static Path packageRoot(Path filepath) throws ProjectException;`~~
~~`public static boolean isBallerinaStandaloneFile(Path filepath);`~~
~~`public static boolean isBallerinaSourceFile(Path filepath);`~~

~~`public static Path findPackageRoot(Path filepath) throws ProjectException;`~~
~~`public static boolean isBallerinaStandaloneFile(Path file);`~~
 Fixes #27273

_**io.ballerina.projects.directory.Project**_
`public abstract DocumentId documentId(Path file);`
Fixes #27274

* Removes the following method:
_**io.ballerina.projects.directory.ProjectLoader**_
`public static Optional<DocumentId> getDocumentId(Path documentFilePath, BuildProject project);`

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/ProjectPaths.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
+package io.ballerina.projects.directory;
```
Comment (hemikak) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/ProjectPaths.java:1]: License header missing.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/BuildProject.java [hunk 2]
```diff
@@ -111,4 +114,38 @@ private BuildProject(ProjectEnvironmentBuilder environmentBuilder, Path projectP
...
         return Optional.empty();
     }
+
+    @Override
+    public DocumentId documentId(Path file) {
+        ProjectPaths.findPackageRoot(file);
+        Path parent = Optional.of(file.toAbsolutePath().getParent()).get();
+        for (ModuleId moduleId : this.currentPackage().moduleIds()) {
+            String moduleDirName;
+            if (moduleId.moduleName().contains(DOT)) {
```
Comment (sameerajayasoma) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/BuildProject.java:124]: It is not correct to get the module name from the `moduleId`. Use `moduleId` to get the ModuleContext.

What's inside moduleId is just for debugging purposes.

Comment (sameerajayasoma) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/BuildProject.java:124]: `ModuleId.moduleName()` is public method which is not correct. Please remove that as well.

**Review by sameerajayasoma — CHANGES_REQUESTED** [review]

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/BuildProject.java (hunk 2 above)
Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/BuildProject.java:124]: ModuleContext is not visible to the BuildProject. What should we do in that case?

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/ProjectPaths.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+public class ProjectPaths {
+
+    /**
+     * Finds the root directory of a Ballerina package using the filepath provided.
+     *
+     * @param filepath ballerina file that belongs to a package
+     * @return path to the package root directory
+     * @throws ProjectException if the provided path is invalid or if it is a standalone ballerina file
+     */
+    public static Path findPackageRoot(Path filepath) throws ProjectException {
```
[the remark this answers is not in the document]
Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/ProjectPaths.java:44]: Yes, the discussed APIs are the ones you mentioned. +1, to add other files. When we support resources, should we support multiple directory levels?

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/ProjectPaths.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+import org.wso2.ballerinalang.compiler.util.ProjectDirConstants;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Optional;
+
+/**
+ * Consists of static methods that may be used to obtain {@link Project}
+ * information using file paths.
+ */
```
[the remark this answers is not in the document]
Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/ProjectPaths.java]: Added in 1594be0

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/ProjectPaths.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/ProjectPaths.java]: Moved in 1594be0

Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/ProjectPaths.java]: Added in 1594be0

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/BuildProject.java (hunk 2 above)
Comment (sameerajayasoma) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/BuildProject.java:125]: We need to restructure the Java packages in `io.ballerina.projects` to make it work. We can move whatever in the `directory` package to the `io.ballerina.projects` package.

Comment (sameerajayasoma) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/BuildProject.java]: Let's go ahead with the current implementation for now. We need to refactor the packages to fix this properly

**Comment by rasika** [comment]
@azinneera any update on this?

**Review by sameerajayasoma — APPROVED** [review]
Had a chat with @azinneera and reviewed the public API methods

**Pull request merged**